### ISSUE TITLE
Add repeat options

### ISSUE THREAD

**Issue by rabbiveesh** [GrantStreetGroup/p5-App-Lrrr#1]
It would be nice to be able to use this tool to automate running tests on change of files. I have something I wrote myself that does this, that it'll rerun tests on change until it gets a 0 return value from the call to system.

I wouldn't mind implementing it for you if this is something that anyone else would use.

It would be in the form of a command line option which would tell it to only rerun until it gets a 0 exit code.

**Comment by preaction** [comment]
I was thinking something similar: A flag that would cause it to run the command and not automatically restart it until a file changes, and maybe another flag that would tell it to retry on errors. Combined with [`prove --state=failed,save`](https://perldoc.perl.org/prove.html#--state) it would be quite interesting :)

So, yes, I'd certainly entertain a PR with something like that :+1:

**Comment by rabbiveesh** [comment]
eh, what do you think the flag should be? Should we assume that there'll be multiple modes, so it'll be a -m flag with named modes?
Also, do we care for compatibility on all platforms? It seems like i can rely on the exit code from the call to `wait` to determine success, but I don't know how to do that reliably on all platforms.

**Comment by preaction** [comment]
[perldoc perlfork](https://perldoc.perl.org/perlfork.html) says that the fork emulation on Windows returns a pseudo-PID that can be passed to `wait`/`waitpid`, so that should work fine. But otherwise, this project should probably have about the [same portability claims for Windows as Mojolicious](https://mojolicious.org/perldoc/Mojolicious/Guides/FAQ#How-well-is-Windows-supported-by-Mojolicious): We'll try to keep it installable, accept patches (with tests) to fix problems on Windows, and try not to cause regressions that break Windows, but we don't do work on Windows, so we can't claim perfect support.

Either way, I'd say we fulfill the current need first and let the future problems come up in the future. No need to spend weeks/months making it perfect when it can be useful quickly :).

As for what the flags should be, that's an excellent question. I think we should probably avoid redefining options that `morbo` uses as anything other than what `morbo` uses them for (to avoid confusion). Maybe `-t|--test` to enable "test mode" which runs the command once when any watched file changes.

Does it need an option to stop when the command exits `0` (or some other exit code)? If I were using this to run tests, I'd want it to keep running tests until I told it to stop, even if they're all passing.

**Comment by rabbiveesh** [comment]
In terms of running the file once when things change, that's actually the
default behaviour. The option I was looking to add was to stop when it gets
an exit code of 0, b/c sometimes I'm trying to focus on a subset of tests
to get them passing, and I wanna only run them, then afterwards I'll run
the full suite.

**Comment by preaction** [comment]
Okay, so I was thinking that `-t` would run the command and then not run it again until a file changed, no matter what the exit code from the command was. It would keep re-running the command as long as watched files change, and using `Ctrl-C` would stop it. Like:

```
$ lrrr -t -- prove -lrv t
... verbose test output here ...
# `prove -lrv t` exit: 1
... time passes ...
# lib/My/App.pm changed, running `prove -lrv t`
... verbose test output here ...
# `prove -lrv t` exit: 0
... time passes ...
# t/newtest.t changed, running `prove -lrv t`
... verbose test output here ...
^C
$ 
```

Is that what you're imagining? Or is it something more elaborate / intelligent? Maybe a patch will make things more clear 😀.

I might be confused because `prove`'s exit code is 0 for all tests passing and non-zero for any test failing. So, if all the tests pass, do you want `lrrr` to automatically stop trying to run the tests? Or would it be okay if you have to `Ctrl-C` to stop `lrrr` from running?

**Comment by rabbiveesh** [comment]
Sorry if this wasn't clear.
What you're imagining the -t options would do is actually what it does now.
It doesn't try to restart the command when it finishes.

What I'm imagining is rerunning the command on a file change after an error
(non-zero exit code), and then stopping the rerunning on success. For the
use case I mentioned above, focusing on a few failing tests.

**Comment by preaction** [comment]
Ah, okay. Perhaps `-u|--until`? Something like `lrrr --test --until exit=0 -- prove ...`

**Comment by rabbiveesh** [comment]
I like the --until.
Though I still don't understand why you would need a --test flag. But I
think I have enough of a start to submit a patch.

**Comment by rabbiveesh** [comment]
Hey, sorry it's been a while without a contribution.

What do you think about a --test flag which waits for the program to exit before restarting it. That's definitely something you might want when running tests.

**Comment by preaction** [comment]
No worries, we're not on a timeline here 😀.

Yes, you're right, that also needs to be part of the `--test` flag. So, I think the potential cases are:

* Current behavior
    1. `lrrr prove`: `lrrr` starts up, runs the `prove` command, no files are changed, the `prove` command exits, and `lrrr` starts the `prove` command again
    2. `lrrr prove`: `lrrr` starts up, runs the `prove` command, a file changes while running the command, `lrrr` kills the running `prove` command, `lrrr` runs the `prove` command again
* New behavior to add
    1. `lrrr -t prove`: `lrrr` starts up, runs the `prove` command, `prove` exits zero, then `lrrr` waits for a change to a file before running `prove` again
    2. `lrrr -t prove`: `lrrr` starts up, runs the `prove` command, `prove` exits non-zero, `lrrr` reports that `prove` exited non-zero, then `lrrr` waits for a change to a file before running `prove` again
    3. `lrrr -t prove`: `lrrr` starts up, runs the `prove` command. While the `prove` command is running, a file changes. `lrrr` notices this and waits until the `prove` command is completed before running `prove` again.

So those are the differences between the two modes:

1. In `--test` mode, `lrrr` will not automatically restart the command when it exits
2. In `--test` mode, `lrrr` will not kill the command if any watched file is changed while the command is running
3. In `--test` mode, `lrrr` will start the command again if any watched file has changed since the last time the command was started

Does that cover all the cases, or am I still missing something?

**Comment by afresh1** [comment]
I would like an option for "find flaky test" where `lrrr --something prove` will run prove and then re-run it if it exits zero.

**Comment by rabbiveesh** [comment]
Two things:
@preaction `lrrr` currently does not restart the program it's running when it ends. If that's the intended behavior without a `--test` flag, then we need another ticket about that bug.
@afresh1 an exit of zero means the test was succesful. Do you mean a non-zero exit, and then that would be the `--until` flag we were discussing, or maybe I don't know what you mean by "flaky".

**Comment by afresh1** [comment]
Yes, that would be the previously discussed `--until` flag, although the interface there seems a bit confusing in the example above. 

To clarify, by flaky I mean a test that only fails sometimes.  A common example is when two time dependent things normally happen in the same second, but sometimes the test overlaps a second and so they no longer match.  Being able to add diagnostics and then re-run until it fails is super useful in that case.

**Comment by rabbiveesh** [comment]
Let's open up another ticket to discuss the interface for that. It
shouldn't be hard to implement. I just need to figure out how to test this
stuff.

**Comment by preaction** [comment]
@rabbiveesh: Yes, it currently does not do that. I'll create a new ticket for that bit (#3)
